examples/upstream.rs

```diff
@@ -0,0 +1,325 @@
+        offset: 0,
+        post: std::ptr::null_mut(),
+    },
+    ngx_null_command!(),
+];
+
+ngx_modules!(ngx_http_upstream_custom_module);
+
+#[no_mangle]
+pub static mut ngx_http_upstream_custom_module: ngx_module_t = ngx_module_t {
+    ctx_index: ngx_uint_t::max_value(),
+    index: ngx_uint_t::max_value(),
+    name: std::ptr::null_mut(),
+    spare0: 0,
+    spare1: 0,
+    version: nginx_version as ngx_uint_t,
+    signature: NGX_RS_MODULE_SIGNATURE.as_ptr() as *const c_char,
+
+    ctx: &ngx_http_upstream_custom_ctx as *const _ as *mut _,
+    commands: unsafe { &ngx_http_upstream_custom_commands[0] as *const _ as *mut _ },
+    type_: NGX_HTTP_MODULE as ngx_uint_t,
+
+    init_master: None,
+    init_module: None,
+    init_process: None,
+    init_thread: None,
+    exit_thread: None,
+    exit_process: None,
+    exit_master: None,
+
+    spare_hook0: 0,
+    spare_hook1: 0,
+    spare_hook2: 0,
+    spare_hook3: 0,
+    spare_hook4: 0,
+    spare_hook5: 0,
+    spare_hook6: 0,
+    spare_hook7: 0,
+};
+
+http_upstream_peer_init!(
+    init_custom_peer,
+    |request: &mut Request, us: *mut ngx_http_upstream_srv_conf_t| {
+        ngx_log_debug_http!(request, "custom init peer");
+
+        let mut hcpd = request.pool().alloc_type::<UpstreamPeerData>();
+        if hcpd.is_null() {
+            return Status::NGX_ERROR;
+        }
+
+        // FIXME make this a convenience macro?
+        let hccf: *const SrvConfig = (*us)
+            .srv_conf
+            .offset(ngx_http_upstream_custom_module.ctx_index as isize)
+            as *const SrvConfig;
+
+        // FIXME method or macro?
+        // Casting from Rust types to C function pointers might be better as macros for
+        // original_init_peer, free, and get.
+        //
+        // Casting to ngx_http_request_t might also benefit. Alternatively a trait which config and
+        // upstream data use may work as well.
+        let original_init_peer: unsafe extern "C" fn(
+            *mut ngx_http_request_t,
+            *mut ngx_http_upstream_srv_conf_t,
+        ) -> ngx_int_t = unsafe { mem::transmute((*hccf).original_init_peer) };
+
+        {
+            let r: *mut ngx_http_request_t = unsafe { mem::transmute(&request) };
+            if original_init_peer(r, us) != Status::NGX_OK.into() {
+                return Status::NGX_ERROR;
+            }
+        }
+
+        let upstream_ptr = request.upstream();
+
+        (*hcpd).conf = Some(hccf);
+        (*hcpd).upstream = upstream_ptr;
+        (*hcpd).data = (*upstream_ptr).peer.data;
+        (*hcpd).client_connection = request.connection();
+        (*hcpd).original_get_peer = (*upstream_ptr).peer.get;
+        (*hcpd).original_free_peer = (*upstream_ptr).peer.free;
+
+        (*upstream_ptr).peer.data = hcpd as *mut c_void;
+        (*upstream_ptr).peer.get = Some(ngx_http_upstream_get_custom_peer);
+        (*upstream_ptr).peer.free = Some(ngx_http_upstream_free_custom_peer);
```

Comment:
NIT: potentially it is a good use case to "wrap upstream" and see what we can do with a new "wrapped" upstream rust type... not insisting on the current PR.

Reply:
I think that's preferable, to wrap it. I've alluded a little to that in the FIXME's. I'm still thinking through how that should look - it's a little complicated to me ATM.

There are a significant number of structures and functions that accept an ngx_http_upstream_t as an argument.
- upstream peers
- upstream nexts
- upstream headers
- upstream locals
- upstream servers
- upstream state

Reply:
I was leaning towards From trait, but need to experiment a little.

I'm hoping a wrapped upstream can help with the `get` and `free` peer functions. If that requires another peer wrapper I need to take that into consideration, but I'd like to hide the Rust -> C func-ptr transmute casts if possible, a macro could possibly hide the boiler plate too perhaps. Once again, need to experiment a little more.